### PR TITLE
Reuse HttpClient and fix the wrong TLS and Proxy settings during MinioClient creation

### PR DESCRIPTION
**Pull request by ebozduman**
Socket/Memory leak issue was addressed before, but there was a case where we were not using HttpHandler when MinioClient is created with no args.
Also adds a functional test for socket/memory leak issue.

### CODE REVIEW

**Comment by harshavardhana** [comment]
> Running `dotnet regitlint --fail-on-diff` for the build issue

You need fix, your formatting is all wrong for a .Net project source. Also you can't remove TLS skip for verification. Or you would have to trust the certs in CI/CD

**Comment by ebozduman** [comment]
OK. I'll add TLS back.

About formatting:
I use `VisualStudio` as my editor and I've disabled `format on save` setting, but it still forces format changes. I'll check it in again. I'll also make sure I manually run `dotnet regitlint` before pushing the changes going forward.

File: Minio.Functional.Tests/Program.cs [hunk 1]
```diff
@@ -68,7 +66,7 @@ public static void Main(string[] args)
                 .WithHttpClient(httpClient)
                 .Build();
         else
-            minioClient = new MinioClient()
+            minioClient = new MinioClient(httpClient)
                 .WithCredentials(accessKey, secretKey)
```
Comment (harshavardhana) [inline Minio.Functional.Tests/Program.cs:70]: You should use WithHttpClient to follow builder pattern.

File: Minio/MinioClient.cs [hunk 2]
```diff
@@ -437,6 +441,11 @@ public MinioClient WithRetryPolicy(RetryPolicyHandlingDelegate retryPolicyHandle
     /// <returns></returns>
     public MinioClient WithHttpClient(HttpClient httpClient)
     {
+        if (HTTPClient != null)
```
Comment (harshavardhana) [inline Minio/MinioClient.cs:444]: This function should throw an exception for httpClient bring null which is the input as InvalidArgumentException

Comment (harshavardhana) [reply Minio/MinioClient.cs:444]: Also this If condition is not needed unless we kind of support all kinds of constructors, we should perhaps remove old style constructors (deprecated ones), move to 5.0.0 and drop all of their support.

**Review by harshavardhana — APPROVED** [review]

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Some reviews

File: Minio.Functional.Tests/Program.cs (hunk 1 above)
Comment (ebozduman) [reply Minio.Functional.Tests/Program.cs:70]: Ok

Comment (ebozduman) [reply Minio.Functional.Tests/Program.cs:70]: Done.

**Comment by ebozduman** [comment]
Implemented review comments and removed the socket/memory test per team decision.

Marked the parameterized construction method, `MinioClient(HttpClient httpClient)`, obsolete. If user has this method used in his/her code, then s/he will see a message similar to the following warning message:
```
.../Minio.Functional.Tests/Program.cs(69,27): warning CS0618: 'MinioClient.MinioClient(HttpClient)' is obsolete: 
'Use MinioClient() and Builder method .WithHttpClient(httpClient)' [.../Minio.Functional.Tests/Minio.Functional.Tests.csproj]
```

File: Minio/MinioClient.cs (hunk 2 above)
Comment (ebozduman) [reply Minio/MinioClient.cs]: Done.

File: Minio/MinioClient.cs [hunk 3]
```diff
@@ -87,13 +87,18 @@ public MinioClient()
         Region = "";
         SessionToken = "";
         Provider = null;
-        HTTPClient = new HttpClient();
+        var clientHandler = new HttpClientHandler();
+        clientHandler.ServerCertificateCustomValidationCallback =
+            (sender, cert, chain, sslPolicyErrors) => { return true; };
+        clientHandler.UseProxy = false;
```
Comment (harshavardhana) [inline Minio/MinioClient.cs:93]: You can't do this either - we should honor local system proxies.

File: Minio/MinioClient.cs [hunk 4]
```diff
@@ -87,13 +87,18 @@ public MinioClient()
         Region = "";
         SessionToken = "";
         Provider = null;
-        HTTPClient = new HttpClient();
+        var clientHandler = new HttpClientHandler();
+        clientHandler.ServerCertificateCustomValidationCallback =
+            (sender, cert, chain, sslPolicyErrors) => { return true; };
```
Comment (harshavardhana) [inline Minio/MinioClient.cs:92]: if you do this then SSL verification will be turned-off in the SDK we cannot have an insecure TLS enforcing HTTP client, we need to make sure by default we ask for trust and security. 

I am not why this code is needed.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: Minio/MinioClient.cs (hunk 3 above)
Comment (ebozduman) [reply Minio/MinioClient.cs:93]: OK.
I'll remove it.

File: Minio/MinioClient.cs (hunk 4 above)
Comment (ebozduman) [reply Minio/MinioClient.cs:92]: I'll remove it.

Comment (ebozduman) [reply Minio/MinioClient.cs]: Removed

**Comment by harshavardhana** [comment]
I am not sure what is the bug fix here? @ebozduman - can you please explain?

**Comment by ebozduman** [comment]
Sorry for the confusion.
Things changed in time and the title and what has changed diverged.

TLDR: This PR will make the last 3 test cases reuse the existing HttpClient instance and fix the wrong TLS and Proxy settings when creating an instance of MinioClient.

Some more explanations:
After the original socket leakage problem is addressed (too many HttpClient instances were created instead of reusing the existing one), I wanted to create a test for it and add some extra clean up for the 3 tests that were still creating their own HttpClient (and MinioClient) instances. Those 3 tests depended on `DownloadObjectAsync` and `UploadObjectAsync` methods. 

In the meantime, I discussed with Bala and like the other SDKs, we've decided a functional test for socket leakage is not necessary.

I also got the reviews from you for the unnecessary HttpHandler creation which was turning off TLS and the review about the proxy setting that causes not honoring local system proxies.

So, this PR will make the last 3 test cases to reuse the existing HttpClient instance and fix the wrong TLS and Proxy settings.
I'll change the title to reflect the correct definition of these changes.

**Review by harshavardhana — APPROVED** [review]
